JointConsensus/JointConsensus.tla

```diff
@@ -0,0 +1,536 @@
+--------------------------------- MODULE JointConsensus ---------------------------------
+\* This is the formal specification for the Raft consensus algorithm.
+\*
+\* Copyright 2014 Diego Ongaro.
+\* This work is licensed under the Creative Commons Attribution-4.0
+\* International License https://creativecommons.org/licenses/by/4.0/
+
+EXTENDS Naturals, FiniteSets, Sequences, TLC
+
+\* The set of server IDs
+CONSTANTS InitialServers, NewServers
+
+CONSTANTS ValueEntry,ConfigEntry
+
+\* Servers states.
+CONSTANTS Follower, Candidate, Leader
+
+\* A reserved value.
+CONSTANTS Nil
+
+\* Message types:
+CONSTANTS RequestVoteRequest, RequestVoteResponse,
+          AppendEntriesRequest, AppendEntriesResponse
+
+
+\* Maximum number of client requests
+CONSTANTS MaxClientRequests
+
+----
+\* Global variables
+
+\* A bag of records representing requests and responses sent from one server
+\* to another. TLAPS doesn't support the Bags module, so this is a function
+\* mapping Message to Nat.
+VARIABLE messages
+
+\* A history variable used in the proof. This would not be present in an
+\* implementation.
+\* Keeps track of successful elections, including the initial logs of the
+\* leader and voters' logs. Set of functions containing various things about
+\* successful elections (see BecomeLeader).
+VARIABLE elections
+
+\* Servers list all known servers.
+VARIABLE allServers
+
+\* The members of the server at that time.
+VARIABLE voters
+VARIABLE nextVoters
```

Comment:
Do we need to consider learners? It's possible to promote a learner, remove them, or add them during Joint Consensus.

Reply:
Maybe later. It will increase the state space.